internal/clients/authentication.go

```diff
@@ -0,0 +1,65 @@
+package clients
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/RHEnVision/provisioning-backend/internal/ctxval"
+	"github.com/RHEnVision/provisioning-backend/internal/models"
+)
+
+type Authentication interface {
+	// Is checks if Authentication is of a given provider type
+	Is(providerType models.ProviderType) bool
+
+	// MustBe returns nil, if authentication is of given time. Otherwise, returns an error.
+	MustBe(providerType models.ProviderType) error
+
+	// String returns authentication string (ARN, Subscription UUID, Project-ID...)
+	String() string
+}
+
+type auth struct {
+	ProviderType models.ProviderType `json:"type"`
+	Payload      string              `json:"payload"`
+}
+
+func NewAuthentication(str string, provType models.ProviderType) Authentication {
+	auth := auth{
+		Payload:      str,
+		ProviderType: provType,
+	}
+	return &auth
+}
+
+func NewAuthenticationFromSourceAuthType(ctx context.Context, str, authType string) Authentication {
+	auth := auth{Payload: str}
+	switch authType {
+	case "provisioning-arn":
+		auth.ProviderType = models.ProviderTypeAWS
+	case "provisioning_lighthouse_subscription_id":
+		auth.ProviderType = models.ProviderTypeAzure
+	case "provisioning_project_id":
+		auth.ProviderType = models.ProviderTypeGCP
+	default:
+		ctxval.Logger(ctx).Warn().Msgf("Unknown auth type returned from sources: %s", authType)
+		auth.ProviderType = models.ProviderTypeUnknown
+	}
+	return &auth
+}
+
+func (auth *auth) Is(providerType models.ProviderType) bool {
+	return auth.ProviderType == providerType
+}
+
+func (auth *auth) MustBe(providerType models.ProviderType) error {
+	if !auth.Is(providerType) {
+		return fmt.Errorf("%w: %s", UnknownAuthenticationTypeErr, auth.ProviderType)
```

Comment:
I guess you need to change it to %v

Reply:
Right, weird that lint does not warn about this. Also provider type did not implement `String()` so it was also an issue.

Reply:
Oh actually no, the linter was only complaining about the 2nd parameter, the first one needs to be wrapped error.